feat: implement WordPress native search endpoint

lucymtc opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes https://github.com/10up/headstartwp/issues/546

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @username, @username2, ...


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

lucymtc committed:
feat: prepare embedable post links in search results

lucymtc commented on wp/headless-wp/includes/classes/Search/PostSearchHandler.php:
```diff
@@ -0,0 +1,68 @@
+
+		// Reduce the search response size.
+		if ( ! empty( $links['self'] ) ) {
+			$links['self']['embeddable'] = false;
```
I'm not sure of the implication on removing 'self' from the _embedded data, thought of reducing the size of the payload. But maybe we can leave it as true, as developers can specify what is needed in _embed param. The _embed param accepts bool or array of strings for what to add in `_embedded`

lucymtc committed:
feat: add WP native search fetch strategy

lucymtc committed:
feat: add changeset

lucymtc committed:
Merge branch 'develop' into feature/wp-search-endpoint

lucymtc commented on packages/core/src/data/strategies/SearchNativeFetchStrategy.ts:
```diff
@@ -0,0 +1,155 @@
+ * @category Data Fetching
+ */
+export class SearchNativeFetchStrategy<
+	T extends PostEntity = PostEntity,
```
@nicholasio everything seems to work fine with these changes, but I am not sure if I need to create a new SearchPostEntity as the data is not the same structure as a post entity in other WP endpoints. But I am not quite understanding why there are no errors

nicholasio replied:
Yeah the types will end up wrong and those using TS will notice it. We do need a new `SearchEntitty` the result can be either be a post or a term object.

lucymtc replied:
updated

nicholasio committed:
Merge branch 'develop' into feature/wp-search-endpoint

nicholasio commented on packages/core/src/data/strategies/SearchNativeFetchStrategy.ts:
```diff
@@ -0,0 +1,155 @@
+		// Request SEO data.
+		try {
+			const wpUrl = getWPUrl().replace(/\/$/, ''); // Ensure no double slash in url param
+			const localeParam = this.locale ? `&lang=${this.locale}` : '';
```
Does Yoast use this to correctly return the metadata for localized sites?

lucymtc replied:
Yes, needs it for the title, but i need to fix the fact that search isn't picking up the lang

nicholasio commented on packages/core/src/data/strategies/SearchNativeFetchStrategy.ts:
```diff
@@ -0,0 +1,155 @@
+		const queriedObject: QueriedObject = {
+			search: {
+				searchedValue: params.search ?? '',
+				type: 'post',
```
`type` can be post, term, post-format

and subtype post, page, category, post_tag

https://developer.wordpress.org/rest-api/reference/search-results/#schema

We should be able to infer that directly from `params.type`.

lucymtc replied:
@nicholasio subtype can be any custom taxonomy or any custom post type

lucymtc replied:
updated the type

nicholasio commented on wp/headless-wp/includes/classes/Plugin.php:
```diff
@@ -48,6 +48,9 @@ public function register() {
 		$gutenberg->register();
+
+		$search = new Search\Search();
+		$search->register();
```
It looks like your customization ended up removing the "self" link, so when `_embed` is passed it doesn't embed the full post object anymore.

See; https://js1.10up.com/wp-json/wp/v2/search?_embed here it embeds "self" with the full post object.

Perhaps you just need to add the link for the post/term itself?

lucymtc replied:
added back self

nicholasio commented on wp/headless-wp/includes/classes/Search/PostSearchHandler.php:
```diff
@@ -0,0 +1,83 @@
+	 */
+	public function prepare_item_links( $id ) {
+
+		$post  = get_post( $id );
```
The item might not always be a post. It's possible to search for terms: https://js1.10up.com/wp-json/wp/v2/search?type=term

It would be better to check if it's actually a post first.

lucymtc replied:
added check to make sure it's a post

nicholasio reviewed (changes requested):
Nice Job! This is looks really good. Just a few more things and it should be good to go.

lucymtc committed:
fix: add search entities for posts and terms

lucymtc commented:
@nicholasio I've addressed your comments, I still need to do couple of things.
1. Lang in search is not working properly. Not even when fetching with the current useSearch strategy, I need to look into it. 
2. The Page number in title still needs some fixing

lucymtc committed:
fix: page number in search title

lucymtc committed:
fix: lang param in yoast get_head

lucymtc committed:
fix: title extra spaces

lucymtc commented:
@nicholasio I fixed the page number issue. For the lang in search, I think this is an issue specific to the environment. The search endpoint for search in localhost:8888 with lang param it gets ignored. but when I test on other wp setup with polylang the lang filters the result as expected.

nicholasio commented on packages/core/src/data/types/entities.ts:
```diff
@@ -644,6 +644,63 @@ export interface SearchEntity extends Entity {
+	terms?: Record<string, TermEntity[]>;
+
+	_embedded: {
+		author: AuthorEntity[];
```
doesn't it need a _self to PostEntity here?

lucymtc replied:
added

nicholasio commented on packages/core/src/react/hooks/useFetchSearchNative.ts:
```diff
@@ -0,0 +1,86 @@
+
+	const { result, pageInfo, queriedObject } = data;
+
+	const posts = result.map((post) => {
```
I think we'd want to name `posts` something else since it can be used to also search terms.

Alternativelly we keep `posts` if searching for post and `terms` when searching for terms.

lucymtc replied:
changed to searchResults as we also have post-format type

lucymtc committed:
fix: add self to  post search entity and update  data naming

lucymtc committed:
Merge branch 'develop' into feature/wp-search-endpoint

lucymtc committed:
fix: update yoast seo plugin with search fix

nicholasio commented on packages/core/src/data/types/entities.ts:
```diff
@@ -644,6 +644,64 @@ export interface SearchEntity extends Entity {
+ */
+export interface TermSearchEntity extends SearchObjectEntity {
+	_embedded: {
+		self: Array<TermEntity[]>;
```
Is this truly an array of array? Shouldn't it just be `TermEntity[]`

lucymtc replied:
no it's not, updated

nicholasio commented on packages/core/src/data/types/entities.ts:
```diff
@@ -644,6 +644,64 @@ export interface SearchEntity extends Entity {
+	_embedded: {
+		author: AuthorEntity[];
+		'wp:term': Array<TermEntity[]>;
+		self: Array<PostEntity[]>;
```
I think this should be just `PostEntity[]` instead of an array of array.

lucymtc replied:
you right, good catch thanks

lucymtc committed:
fix: search post and term self types

lucymtc committed:
fix: search post and term self types

nicholasio committed:
Merge branch 'feature/wp-search-endpoint' of github.com:10up/headless into feature/wp-search-endpoint

nicholasio commented:
FYI: Somehow I managed to messed up the previous commit that adds some basic tests.

nicholasio committed:
test: fix tests

nicholasio committed:
test: add tests for searching terms

nicholasio committed:
refactor: add support for passing subtype as array of string

nicholasio committed:
docs: useSearchNative

nicholasio reviewed (approved):
Excellent work here @lucymtc. I've added some tests and docs and this is now ready to be merged & released.
